Mathlib/RingTheory/KrullDimension.lean

```diff
@@ -0,0 +1,113 @@
+prime ideals ordered by inclusion.
+
+
+## Main definitions and results
+
+* `ringKrullDim`: the ring theoretic Krull dimension of a commutative ring.
+
+* `ringKrullDim.eq_topologicalKrullDim`: the ring theoretic Krull dimension of a commutative ring
+is equal to the topological Krull dimension of its prime spectrum.
+
+* `ringKrullDim.eq_zero_of_field`: the Krull dimension of a field is zero.
+-/
+
+/--
+The ring theoretic Krull dimension is the Krull dimension of its spectrum ordered by inclusion.
+-/
+noncomputable abbrev ringKrullDim (R : Type*) [CommRing R] : WithBot (WithTop ℕ) :=
+  krullDim (PrimeSpectrum R)
+
+namespace ringKrullDim
+
+lemma eq_bot_of_subsingleton (R : Type*) [CommRing R] [Subsingleton R] :
+    ringKrullDim R = ⊥ :=
+  krullDim_eq_bot_of_isEmpty
+
+lemma nonneg_of_nontrivial (R : Type*) [CommRing R] [Nontrivial R] :
+    0 ≤ ringKrullDim R :=
+  krullDim_nonneg_of_nonempty
+
+theorem eq_topologicalKrullDim (R : Type*) [CommRing R] :
+    ringKrullDim R = topologicalKrullDim (PrimeSpectrum R) :=
+  Eq.symm $ krullDim_orderDual.symm.trans $ krullDim_eq_of_orderIso
+  (PrimeSpectrum.pointsEquivIrreducibleCloseds R).symm
+
+/-- If `f : R →+* S` is surjective, then `ringKrullDim S ≤ ringKrullDim R`. -/
+theorem le_of_surjective {R S : Type*} [CommRing R] [CommRing S] (f : R →+* S)
+    (hf : Function.Surjective f) : ringKrullDim S ≤ ringKrullDim R :=
+  krullDim_le_of_strictMono (PrimeSpectrum.comap f) (Monotone.strictMono_of_injective
+    (fun _ _ hab ↦ Ideal.comap_mono hab) (PrimeSpectrum.comap_injective_of_surjective f hf))
+
+/-- If `I` is an ideal of `R`, then `ringKrullDim (R ⧸ I) ≤ ringKrullDim R`. -/
+theorem quotient_le {R : Type*} [CommRing R] (I : Ideal R) :
+    ringKrullDim (R ⧸ I) ≤ ringKrullDim R :=
+  le_of_surjective _ Ideal.Quotient.mk_surjective
+
+/-- If `R` and `S` are isomorphic, then `ringKrullDim R = ringKrullDim S`. -/
+theorem eq_of_ringEquiv {R S : Type*} [CommRing R] [CommRing S] (e : R ≃+* S) :
+    ringKrullDim R = ringKrullDim S :=
+  le_antisymm (le_of_surjective e.symm e.symm.surjective) (le_of_surjective e e.surjective)
+
+section DimensionZero
+
+variable {R : Type _} [CommRing R]
+
+theorem _root_.Ideal.IsPrime.isMaximal_of_ringKrullDim_eq_zero
+    {I : Ideal R} (hI : I.IsPrime) (hdim : ringKrullDim R = 0) :
+    I.IsMaximal := by
+  obtain ⟨J, hJ1, hJ2⟩ := I.exists_le_maximal hI.1
+  by_cases hIJ : I = J
+  · subst hIJ; exact hJ1
+  · let x : LTSeries (PrimeSpectrum R) := ⟨1, ![⟨I, hI⟩, ⟨J, hJ1.isPrime⟩],
+      fun i ↦ by fin_cases i; exact ⟨hJ2, fun r ↦ hIJ <| Eq.symm <| hJ1.eq_of_le hI.1 r⟩⟩
+    have : 1 ≤ ringKrullDim R :=
+      le_iSup (α := WithBot (WithTop ℕ)) (f := fun x : LTSeries _ ↦ x.length) x
+    rw [hdim] at this
+    norm_num at this
+
+theorem _root_.Ideal.IsPrime.mem_minimalPrimes_of_ringKrullDim_eq_zero
+    {I : Ideal R} (hI : I.IsPrime) (hdim : ringKrullDim R = 0) :
```

Comment:
There should be a lemma saying `krullDim a = 0 <-> forall x in a, x is maximal (minimal)`, then specialize.

Reply:
I have formalized the following lemmas:
```
lemma mem_maximals_of_krullDim_eq_zero {α : Type*} [Preorder α] :
    krullDim α = 0 → ∀ (a : α), a ∈ maximals (fun x x1 ↦ x ≤ x1) ⊤ := ...

lemma mem_minimals_of_krullDim_eq_zero {α : Type*} [Preorder α] :
    krullDim α = 0 → ∀ (a : α), a ∈ minimals (fun x x1 ↦ x ≤ x1) ⊤ := ...

lemma krullDim_eq_zero_iff_mem_maximals_of_nonempty {α : Type*} [Preorder α] [Nonempty α] :
    krullDim α = 0 ↔ ∀ a : α, a ∈ maximals (· ≤ ·) ⊤ := ...

lemma krullDim_eq_zero_iff_mem_minimals_of_nonempty {α : Type*} [Preorder α] [Nonempty α] :
    krullDim α = 0 ↔ ∀ a : α, a ∈ minimals (· ≤ ·) ⊤ := ...
```
Then I made consequent changes to the results
```
Ideal.IsPrime.isMaximal_of_ringKrullDim_eq_zero

Ideal.IsPrime.mem_minimalPrimes_of_ringKrullDim_eq_zero
```